components/brave_shields/browser/ad_block_subscription_service_manager.cc

```diff
@@ -0,0 +1,459 @@
+    const std::string& url) {
+  base::Optional<base::Value> first_value = base::nullopt;
+
+  base::AutoLock lock(subscription_services_lock_);
+  for (auto it = subscription_services_.begin();
+       it != subscription_services_.end(); it++) {
+    auto info = GetInfo(it->first);
+    if (info && info->enabled) {
+      base::Optional<base::Value> next_value =
+          it->second->UrlCosmeticResources(url);
+      if (first_value) {
+        if (next_value) {
+          MergeResourcesInto(std::move(*next_value), &*first_value, false);
+        }
+      } else {
+        first_value = std::move(next_value);
+      }
+    }
+  }
+
+  return first_value;
+}
+
+base::Optional<base::Value>
+AdBlockSubscriptionServiceManager::HiddenClassIdSelectors(
+    const std::vector<std::string>& classes,
+    const std::vector<std::string>& ids,
+    const std::vector<std::string>& exceptions) {
+  base::Optional<base::Value> first_value = base::nullopt;
+
+  base::AutoLock lock(subscription_services_lock_);
+  for (auto it = subscription_services_.begin();
+       it != subscription_services_.end(); it++) {
+    auto info = GetInfo(it->first);
+    if (info && info->enabled) {
+      base::Optional<base::Value> next_value =
+          it->second->HiddenClassIdSelectors(classes, ids, exceptions);
+      if (first_value && first_value->is_list()) {
+        if (next_value && next_value->is_list()) {
+          for (auto i = next_value->GetList().begin();
+               i < next_value->GetList().end(); i++) {
+            first_value->Append(std::move(*i));
+          }
+        }
+      } else {
+        first_value = std::move(next_value);
+      }
+    }
+  }
+
+  return first_value;
+}
+
+void AdBlockSubscriptionServiceManager::OnListDownloaded(
+    const SubscriptionIdentifier& id) {
+  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
+  auto it = subscription_services_.find(id);
+  if (it == subscription_services_.end())
+    return;
+
+  it->second->OnSuccessfulDownload();
+}
+
+void AdBlockSubscriptionServiceManager::OnListLoaded(
+    const SubscriptionIdentifier& id) {
+  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
+  auto it = subscription_services_.find(id);
+  if (it == subscription_services_.end())
+    return;
+
+  auto info = GetInfo(id);
+  if (!info)
+    return;
+
+  info->last_update_attempt = base::Time::Now();
+  info->last_successful_update_attempt = base::Time::Now();
+  UpdateFilterListPrefs(id, *info);
+
+  // calling this more than once is ok
+  it->second->Start();
+
+  NotifyObserversOfServiceEvent();
+}
+
+void AdBlockSubscriptionServiceManager::OnListDownloadFailure(
+    const SubscriptionIdentifier& id) {
+  auto info = GetInfo(id);
+  if (!info)
+    return;
+
+  info->last_update_attempt = base::Time::Now();
+  UpdateFilterListPrefs(id, *info);
+}
+
+void AdBlockSubscriptionServiceManager::NotifyObserversOfServiceEvent() {
+  for (auto& observer : observers_) {
+    observer.OnServiceUpdateEvent();
```

Comment:
`OnServiceUpdate`? probably `OnListLoaded` makes more sense here. Also the convention for the method name is `NotifyEventName` so `NotifyListLoaded` for `OnListLoaded`